[BUG] NoSuchElementException: None.get

**Describe the bug**
There are frequent (once per couple minutes) errors in the Node log:
```
Oct 03 11:12:48 rock64 waves[10468]: INFO  c.w.s.BlockchainUpdaterImpl - New height: 1733623
Oct 03 11:12:48 rock64 waves[10468]: ERROR c.w.network.HistoryReplier - Error executing task
Oct 03 11:12:48 rock64 waves[10468]: com.google.common.util.concurrent.UncheckedExecutionException: java.util.NoSuchElementException: None.get
Oct 03 11:12:48 rock64 waves[10468]:         at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2050)
Oct 03 11:12:48 rock64 waves[10468]:         at com.google.common.cache.LocalCache.get(LocalCache.java:3952)
Oct 03 11:12:48 rock64 waves[10468]:         at com.google.common.cache.LocalCache.getOrLoad(LocalCache.java:3974)
Oct 03 11:12:48 rock64 waves[10468]:         at com.google.common.cache.LocalCache$LocalLoadingCache.get(LocalCache.java:4958)
Oct 03 11:12:48 rock64 waves[10468]:         at com.wavesplatform.network.HistoryReplier.$anonfun$channelRead$7(HistoryReplier.scala:59)
Oct 03 11:12:48 rock64 waves[10468]:         at monix.eval.internal.TaskRunLoop$.startFuture(TaskRunLoop.scala:494)
Oct 03 11:12:48 rock64 waves[10468]:         at monix.eval.Task.runToFutureOpt(Task.scala:585)
Oct 03 11:12:48 rock64 waves[10468]:         at monix.eval.Task.runToFuture(Task.scala:541)
Oct 03 11:12:48 rock64 waves[10468]:         at com.wavesplatform.utils.ScorexLogging$TaskExt.runAsyncLogErr(ScorexLogging.scala:66)
Oct 03 11:12:48 rock64 waves[10468]:         at com.wavesplatform.network.HistoryReplier.channelRead(HistoryReplier.scala:65)
Oct 03 11:12:48 rock64 waves[10468]:         at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
Oct 03 11:12:48 rock64 waves[10468]:         at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
Oct 03 11:12:48 rock64 waves[10468]:         at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
Oct 03 11:12:48 rock64 waves[10468]:         at io.netty.channel.ChannelInboundHandlerAdapter.channelRead(ChannelInboundHandlerAdapter.java:86)
Oct 03 11:12:48 rock64 waves[10468]:         at com.wavesplatform.network.PeerSynchronizer.channelRead(PeerSynchronizer.scala:55)
Oct 03 11:12:48 rock64 waves[10468]:         at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
Oct 03 11:12:48 rock64 waves[10468]:         at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
Oct 03 11:12:48 rock64 waves[10468]:         at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
Oct 03 11:12:48 rock64 waves[10468]:         at io.netty.channel.ChannelInboundHandlerAdapter.channelRead(ChannelInboundHandlerAdapter.java:86)
Oct 03 11:12:48 rock64 waves[10468]:         at com.wavesplatform.network.TrafficLogger.channelRead(TrafficLogger.scala:41)
Oct 03 11:12:48 rock64 waves[10468]:         at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
Oct 03 11:12:48 rock64 waves[10468]:         at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
Oct 03 11:12:48 rock64 waves[10468]:         at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
Oct 03 11:12:48 rock64 waves[10468]:         at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:102)
Oct 03 11:12:48 rock64 waves[10468]:         at io.netty.handler.codec.MessageToMessageCodec.channelRead(MessageToMessageCodec.java:111)
Oct 03 11:12:48 rock64 waves[10468]:         at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
Oct 03 11:12:48 rock64 waves[10468]:         at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
Oct 03 11:12:48 rock64 waves[10468]:         at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
Oct 03 11:12:48 rock64 waves[10468]:         at io.netty.channel.ChannelInboundHandlerAdapter.channelRead(ChannelInboundHandlerAdapter.java:86)
Oct 03 11:12:48 rock64 waves[10468]:         at com.wavesplatform.network.DiscardingHandler.channelRead(DiscardingHandler.scala:18)
Oct 03 11:12:48 rock64 waves[10468]:         at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
Oct 03 11:12:48 rock64 waves[10468]:         at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
Oct 03 11:12:48 rock64 waves[10468]:         at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
Oct 03 11:12:48 rock64 waves[10468]:         at io.netty.channel.ChannelInboundHandlerAdapter.channelRead(ChannelInboundHandlerAdapter.java:86)
Oct 03 11:12:48 rock64 waves[10468]:         at com.wavesplatform.network.TrafficWatcher.channelRead(TrafficWatcher.scala:47)
Oct 03 11:12:48 rock64 waves[10468]:         at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
Oct 03 11:12:48 rock64 waves[10468]:         at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
Oct 03 11:12:48 rock64 waves[10468]:         at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
Oct 03 11:12:48 rock64 waves[10468]:         at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:323)
Oct 03 11:12:48 rock64 waves[10468]:         at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:297)
Oct 03 11:12:48 rock64 waves[10468]:         at io.netty.handler.codec.ByteToMessageCodec.channelRead(ByteToMessageCodec.java:103)
Oct 03 11:12:48 rock64 waves[10468]:         at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
Oct 03 11:12:48 rock64 waves[10468]:         at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
Oct 03 11:12:48 rock64 waves[10468]:         at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
Oct 03 11:12:48 rock64 waves[10468]:         at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:323)
Oct 03 11:12:48 rock64 waves[10468]:         at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:297)
Oct 03 11:12:48 rock64 waves[10468]:         at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
Oct 03 11:12:48 rock64 waves[10468]:         at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
Oct 03 11:12:48 rock64 waves[10468]:         at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
Oct 03 11:12:48 rock64 waves[10468]:         at io.netty.channel.ChannelInboundHandlerAdapter.channelRead(ChannelInboundHandlerAdapter.java:86)
Oct 03 11:12:48 rock64 waves[10468]:         at com.wavesplatform.network.BrokenConnectionDetector.channelRead(BrokenConnectionDetector.scala:17)
Oct 03 11:12:48 rock64 waves[10468]:         at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
Oct 03 11:12:48 rock64 waves[10468]:         at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
Oct 03 11:12:48 rock64 waves[10468]:         at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
Oct 03 11:12:48 rock64 waves[10468]:         at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1408)
Oct 03 11:12:48 rock64 waves[10468]:         at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
Oct 03 11:12:48 rock64 waves[10468]:         at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
Oct 03 11:12:48 rock64 waves[10468]:         at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:930)
Oct 03 11:12:48 rock64 waves[10468]:         at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:163)
Oct 03 11:12:48 rock64 waves[10468]:         at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:677)
Oct 03 11:12:48 rock64 waves[10468]:         at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:612)
Oct 03 11:12:48 rock64 waves[10468]:         at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:529)
Oct 03 11:12:48 rock64 waves[10468]:         at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:491)
Oct 03 11:12:48 rock64 waves[10468]:         at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:905)
Oct 03 11:12:48 rock64 waves[10468]:         at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
Oct 03 11:12:48 rock64 waves[10468]:         at java.lang.Thread.run(Thread.java:748)
Oct 03 11:12:48 rock64 waves[10468]: Caused by: java.util.NoSuchElementException: None.get
Oct 03 11:12:48 rock64 waves[10468]:         at scala.None$.get(Option.scala:529)
Oct 03 11:12:48 rock64 waves[10468]:         at scala.None$.get(Option.scala:527)
Oct 03 11:12:48 rock64 waves[10468]:         at com.wavesplatform.network.HistoryReplier$$anon$1.load(HistoryReplier.scala:27)
Oct 03 11:12:48 rock64 waves[10468]:         at com.wavesplatform.network.HistoryReplier$$anon$1.load(HistoryReplier.scala:24)
Oct 03 11:12:48 rock64 waves[10468]:         at com.google.common.cache.LocalCache$LoadingValueReference.loadFuture(LocalCache.java:3528)
Oct 03 11:12:48 rock64 waves[10468]:         at com.google.common.cache.LocalCache$Segment.loadSync(LocalCache.java:2277)
Oct 03 11:12:48 rock64 waves[10468]:         at com.google.common.cache.LocalCache$Segment.lockedGetOrLoad(LocalCache.java:2154)
Oct 03 11:12:48 rock64 waves[10468]:         at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2044)
Oct 03 11:12:48 rock64 waves[10468]:         ... 65 common frames omitted
Oct 03 11:12:48 rock64 waves[10468]: INFO  c.w.s.BlockchainUpdaterImpl - MicroBlock(5ficwpH... -> Nn5Y3jY..., txs=6) appended
Oct 03 11:12:54 rock64 waves[10468]: INFO  c.w.s.BlockchainUpdaterImpl - MicroBlock(3WTQnZt... -> 5ficwpH..., txs=40) appended
```
**Additional context**
These problems seem to be started several days ago, after Feature 14 approved

--- a/node-it/src/test/scala/com/wavesplatform/it/sync/RollbackSuite.scala
+++ b/node-it/src/test/scala/com/wavesplatform/it/sync/RollbackSuite.scala
@@ -4,7 +4,7 @@ import com.typesafe.config.Config
 import com.wavesplatform.account.KeyPair
 import com.wavesplatform.it.api.SyncHttpApi._
 import com.wavesplatform.it.transactions.NodesFromDocker
-import com.wavesplatform.it.{Node, NodeConfigs, TransferSending}
+import com.wavesplatform.it.{Node, NodeConfigs, ReportingTestName, TransferSending}
 import com.wavesplatform.lang.v2.estimator.ScriptEstimatorV2
 import com.wavesplatform.state.{BooleanDataEntry, IntegerDataEntry, StringDataEntry}
 import com.wavesplatform.transaction.smart.SetScriptTransaction
@@ -16,7 +16,14 @@ import scala.concurrent.Await
 import scala.concurrent.duration._
 import scala.util.Random
 
-class RollbackSuite extends FunSuite with CancelAfterFailure with TransferSending with NodesFromDocker with Matchers with TableDrivenPropertyChecks {
+class RollbackSuite
+    extends FunSuite
+    with CancelAfterFailure
+    with TransferSending
+    with NodesFromDocker
+    with ReportingTestName
+    with Matchers
+    with TableDrivenPropertyChecks {
   override def nodeConfigs: Seq[Config] =
     NodeConfigs.newBuilder
       .overrideBase(_.quorum(0))
@@ -207,7 +214,7 @@ class RollbackSuite extends FunSuite with CancelAfterFailure with TransferSendin
     nodes.waitForHeightArise()
 
     assert(sender.findTransactionInfo(dtx).isDefined)
-    assert(sender.findTransactionInfo(tx).isEmpty)
+    assert(sender.findTransactionInfo(tx).isDefined)
 
   }
 
@@ -216,7 +223,8 @@ class RollbackSuite extends FunSuite with CancelAfterFailure with TransferSendin
       ("num", "name"),
       (1, "1 of N"),
       (nodes.size, "N of N")
-    )) { (num, name) =>
+    )
+  ) { (num, name) =>
     test(s"generate more blocks and resynchronise after rollback $name") {
       val baseHeight = nodes.map(_.height).max + 5
       nodes.waitForHeight(baseHeight)

--- a/node/src/main/scala/com/wavesplatform/Application.scala
+++ b/node/src/main/scala/com/wavesplatform/Application.scala
@@ -270,6 +270,7 @@ class Application(val actorSystem: ActorSystem, val settings: WavesSettings, con
           peerDatabase,
           establishedConnections,
           blockId => Task(blockchainUpdater.removeAfter(blockId)).executeOn(appenderScheduler),
+          discardedTxs => utxStorage.addAndCleanup(discardedTxs),
           allChannels,
           utxStorage,
           miner,

--- a/node/src/main/scala/com/wavesplatform/Application.scala
+++ b/node/src/main/scala/com/wavesplatform/Application.scala
@@ -269,9 +269,18 @@ class Application(val actorSystem: ActorSystem, val settings: WavesSettings, con
           blockchainUpdater,
           peerDatabase,
           establishedConnections,
-          blockId => Task(blockchainUpdater.removeAfter(blockId)).executeOn(appenderScheduler),
-          discardedTxs => utxStorage.addAndCleanup(discardedTxs),
-          allChannels,
+          (blockId, returnTxsToUtx) =>
+            Task(blockchainUpdater.removeAfter(blockId))
+              .executeOn(appenderScheduler)
+              .asyncBoundary
+              .map {
+                case Right(discardedBlocks) =>
+                  allChannels.broadcast(LocalScoreChanged(blockchainUpdater.score))
+                  if (returnTxsToUtx) utxStorage.addAndCleanup(discardedBlocks.view.flatMap(_.transactionData))
+                  miner.scheduleMining()
+                  Right(())
+                case Left(error) => Left(error)
+              },
           utxStorage,
           miner,
           historyReplier,

--- a/node/src/main/scala/com/wavesplatform/http/DebugApiRoute.scala
+++ b/node/src/main/scala/com/wavesplatform/http/DebugApiRoute.scala
@@ -11,14 +11,13 @@ import com.typesafe.config.{ConfigObject, ConfigRenderOptions}
 import com.wavesplatform.account.Address
 import com.wavesplatform.api.http.ApiError.InvalidAddress
 import com.wavesplatform.api.http._
-import com.wavesplatform.block.Block
 import com.wavesplatform.block.Block.BlockId
 import com.wavesplatform.common.state.ByteStr
 import com.wavesplatform.common.utils.Base58
 import com.wavesplatform.crypto
 import com.wavesplatform.lang.ValidationError
 import com.wavesplatform.mining.{Miner, MinerDebugInfo}
-import com.wavesplatform.network.{LocalScoreChanged, PeerDatabase, PeerInfo, _}
+import com.wavesplatform.network.{PeerDatabase, PeerInfo, _}
 import com.wavesplatform.settings.WavesSettings
 import com.wavesplatform.state.diffs.TransactionDiffer
 import com.wavesplatform.state.extensions.Distributions
@@ -31,7 +30,6 @@ import com.wavesplatform.utils.{ScorexLogging, Time}
 import com.wavesplatform.utx.UtxPool
 import com.wavesplatform.wallet.Wallet
 import io.netty.channel.Channel
-import io.netty.channel.group.ChannelGroup
 import io.swagger.annotations._
 import javax.ws.rs.Path
 import monix.eval.{Coeval, Task}
@@ -53,9 +51,7 @@ case class DebugApiRoute(
     ng: NG,
     peerDatabase: PeerDatabase,
     establishedConnections: ConcurrentMap[Channel, PeerInfo],
-    rollbackTask: ByteStr => Task[Either[ValidationError, Seq[Block]]],
-    discardedTxsToUtx: Seq[Transaction] => Unit,
-    allChannels: ChannelGroup,
+    rollbackTask: (ByteStr, Boolean) => Task[Either[ValidationError, Unit]],
     utxStorage: UtxPool,
     miner: Miner with MinerDebugInfo,
     historyReplier: HistoryReplier,
@@ -181,13 +177,8 @@ case class DebugApiRoute(
   private def rollbackToBlock(blockId: ByteStr, returnTransactionsToUtx: Boolean)(
       implicit ec: ExecutionContext
   ): Future[Either[ValidationError, JsObject]] = {
-    rollbackTask(blockId).asyncBoundary
-      .map(_.map { blocks =>
-        allChannels.broadcast(LocalScoreChanged(ng.score))
-        if (returnTransactionsToUtx) discardedTxsToUtx(blocks.view.flatMap(_.transactionData))
-        miner.scheduleMining()
-        Json.obj("BlockId" -> blockId.toString)
-      })
+    rollbackTask(blockId, returnTransactionsToUtx)
+      .map(_ => Right(Json.obj("BlockId" -> blockId.toString)))
       .runAsyncLogErr(Scheduler(ec))
   }
 

--- a/node/src/main/scala/com/wavesplatform/http/DebugApiRoute.scala
+++ b/node/src/main/scala/com/wavesplatform/http/DebugApiRoute.scala
@@ -54,6 +54,7 @@ case class DebugApiRoute(
     peerDatabase: PeerDatabase,
     establishedConnections: ConcurrentMap[Channel, PeerInfo],
     rollbackTask: ByteStr => Task[Either[ValidationError, Seq[Block]]],
+    discardedTxsToUtx: Seq[Transaction] => Unit,
     allChannels: ChannelGroup,
     utxStorage: UtxPool,
     miner: Miner with MinerDebugInfo,
@@ -183,7 +184,7 @@ case class DebugApiRoute(
     rollbackTask(blockId).asyncBoundary
       .map(_.map { blocks =>
         allChannels.broadcast(LocalScoreChanged(ng.score))
-        if (returnTransactionsToUtx) blocks.view.flatMap(_.transactionData).foreach(utxStorage.putIfNew(_))
+        if (returnTransactionsToUtx) discardedTxsToUtx(blocks.view.flatMap(_.transactionData))
         miner.scheduleMining()
         Json.obj("BlockId" -> blockId.toString)
       })

--- a/node/src/main/scala/com/wavesplatform/network/HistoryReplier.scala
+++ b/node/src/main/scala/com/wavesplatform/network/HistoryReplier.scala
@@ -1,6 +1,7 @@
 package com.wavesplatform.network
 
 import com.google.common.cache.{CacheBuilder, CacheLoader}
+import com.google.common.util.concurrent.UncheckedExecutionException
 import com.wavesplatform.common.state.ByteStr
 import com.wavesplatform.network.HistoryReplier._
 import com.wavesplatform.network.MicroBlockSynchronizer.MicroBlockSignature
@@ -18,11 +19,13 @@ class HistoryReplier(ng: NG, settings: SynchronizationSettings, scheduler: Sched
 
   private implicit val s: SchedulerService = scheduler
 
+  // both caches call .get so that NoSuchElementException is thrown and only successfully loaded (micro)blocks are cached
+
   private val knownMicroBlocks = CacheBuilder
     .newBuilder()
     .maximumSize(historyReplierSettings.maxMicroBlockCacheSize)
     .build(new CacheLoader[MicroBlockSignature, Array[Byte]] {
-      override def load(key: MicroBlockSignature) =
+      override def load(key: MicroBlockSignature): Array[Byte] =
         ng.microBlock(key)
           .map(m => MicroBlockResponseSpec.serializeData(MicroBlockResponse(m)))
           .get
@@ -35,40 +38,48 @@ class HistoryReplier(ng: NG, settings: SynchronizationSettings, scheduler: Sched
       override def load(key: ByteStr) = ng.blockBytes(key).get
     })
 
+  private def respondWith(ctx: ChannelHandlerContext, loader: Task[Option[Message]]): Unit =
+    loader.map {
+      case Some(msg) if ctx.channel().isOpen => ctx.writeAndFlush(msg)
+      case _                                 =>
+    }.runAsyncLogErr
+
+  private def handlingNSE[A](f: => A): Option[A] =
+    try Some(f)
+    catch {
+      case uee: UncheckedExecutionException if uee.getCause != null && uee.getCause.isInstanceOf[NoSuchElementException] => None
+    }
+
   override def channelRead(ctx: ChannelHandlerContext, msg: AnyRef): Unit = msg match {
     case GetSignatures(otherSigs) =>
-      Task {
-        val nextIds = otherSigs.view
-          .map(id => id -> ng.blockIdsAfter(id, settings.maxChainLength))
-          .collectFirst { case (parent, Some(ids)) => parent +: ids }
+      respondWith(
+        ctx,
+        Task {
+          val nextIds = otherSigs.view
+            .map(id => id -> ng.blockIdsAfter(id, settings.maxChainLength))
+            .collectFirst { case (parent, Some(ids)) => parent +: ids }
 
-        nextIds match {
-          case Some(extension) =>
-            log.debug(
-              s"${id(ctx)} Got GetSignatures with ${otherSigs.length}, found common parent ${extension.head} and sending total of ${extension.length} signatures")
-            ctx.writeAndFlush(Signatures(extension))
-          case None =>
-            log.debug(s"${id(ctx)} Got GetSignatures with ${otherSigs.length} signatures, but could not find an extension")
+          nextIds match {
+            case Some(extension) =>
+              log.debug(
+                s"${id(ctx)} Got GetSignatures with ${otherSigs.length}, found common parent ${extension.head} and sending total of ${extension.length} signatures"
+              )
+              Some(Signatures(extension))
+            case None =>
+              log.debug(s"${id(ctx)} Got GetSignatures with ${otherSigs.length} signatures, but could not find an extension")
+              None
+          }
         }
-      }.runAsyncLogErr
+      )
 
     case GetBlock(sig) =>
-      Task(knownBlocks.get(sig)).map(bytes => ctx.writeAndFlush(RawBytes(BlockSpec.messageCode, bytes))).logErrDiscardNoSuchElementException.runAsyncLogErr
+      respondWith(ctx, Task(handlingNSE(knownBlocks.get(sig)).map(bytes => RawBytes(BlockSpec.messageCode, bytes))))
 
-    case mbr @ MicroBlockRequest(totalResBlockSig) =>
-      Task(knownMicroBlocks.get(totalResBlockSig))
-        .map { bytes =>
-          ctx.writeAndFlush(RawBytes(MicroBlockResponseSpec.messageCode, bytes))
-          log.trace(id(ctx) + s"Sent MicroBlockResponse(total=${totalResBlockSig.trim})")
-        }
-        .logErrDiscardNoSuchElementException
-        .runAsyncLogErr
+    case MicroBlockRequest(totalResBlockSig) =>
+      respondWith(ctx, Task(handlingNSE(knownMicroBlocks.get(totalResBlockSig)).map(bytes => RawBytes(MicroBlockResponseSpec.messageCode, bytes))))
 
     case _: Handshake =>
-      Task {
-        if (ctx.channel().isOpen)
-          ctx.writeAndFlush(LocalScoreChanged(ng.score))
-      }.runAsyncLogErr
+      respondWith(ctx, Task(Some(LocalScoreChanged(ng.score))))
 
     case _ => super.channelRead(ctx, msg)
   }
@@ -77,7 +88,5 @@ class HistoryReplier(ng: NG, settings: SynchronizationSettings, scheduler: Sched
 }
 
 object HistoryReplier {
-
   case class CacheSizes(blocks: Long, microBlocks: Long)
-
 }

--- a/node/src/main/scala/com/wavesplatform/utils/ScorexLogging.scala
+++ b/node/src/main/scala/com/wavesplatform/utils/ScorexLogging.scala
@@ -1,6 +1,5 @@
 package com.wavesplatform.utils
 
-import com.google.common.util.concurrent.UncheckedExecutionException
 import monix.eval.Task
 import monix.execution.{CancelableFuture, Scheduler}
 import monix.reactive.Observable
@@ -68,20 +67,6 @@ trait ScorexLogging {
     def logErr: Task[A] = {
       t.onErrorHandleWith(ex => {
         log.error(s"Error executing task", ex)
-        Task.raiseError[A](ex)
-      })
-    }
-
-    def logErrDiscardNoSuchElementException: Task[A] = {
-      t.onErrorHandleWith(ex => {
-        ex match {
-          case gex: UncheckedExecutionException =>
-            Option(gex.getCause) match {
-              case Some(nseex: NoSuchElementException) =>
-              case _                                   => log.error(s"Error executing task", ex)
-            }
-          case _ => log.error(s"Error executing task", ex)
-        }
         Task.raiseError[A](ex)
       })
     }

--- a/node/src/test/scala/com/wavesplatform/http/DebugApiRouteSpec.scala
+++ b/node/src/test/scala/com/wavesplatform/http/DebugApiRouteSpec.scala
@@ -11,7 +11,7 @@ class DebugApiRouteSpec extends RouteSpec("/debug") with RestAPISettingsHelper w
   private val wavesSettings = WavesSettings.fromRootConfig(sampleConfig)
   private val configObject  = sampleConfig.root()
   private val route =
-    DebugApiRoute(wavesSettings, ntpTime, null, null, null, null, null, null, null, null, null, null, null, null, null, configObject).route
+    DebugApiRoute(wavesSettings, ntpTime, null, null, null, null, null, null, null, null, null, null, null, null, null, null, configObject).route
 
   routePath("/configInfo") - {
     "requires api-key header" in {

--- a/node/src/test/scala/com/wavesplatform/http/DebugApiRouteSpec.scala
+++ b/node/src/test/scala/com/wavesplatform/http/DebugApiRouteSpec.scala
@@ -11,7 +11,7 @@ class DebugApiRouteSpec extends RouteSpec("/debug") with RestAPISettingsHelper w
   private val wavesSettings = WavesSettings.fromRootConfig(sampleConfig)
   private val configObject  = sampleConfig.root()
   private val route =
-    DebugApiRoute(wavesSettings, ntpTime, null, null, null, null, null, null, null, null, null, null, null, null, null, null, configObject).route
+    DebugApiRoute(wavesSettings, ntpTime, null, null, null, null, null, null, null, null, null, null, null, null, configObject).route
 
   routePath("/configInfo") - {
     "requires api-key header" in {

